sbnci/Modules/TrackValidation/TrackValidation_module.cc

```diff
@@ -0,0 +1,3127 @@
+    }
+  }//Close the loop over TrkidToEnergy map
+  if(likelyParticleID == 0){
+    std::cout<< "The Particle trkid is zero check before continue "<<" Energt is:"<<max_energy<<std::endl;
+  }
+  return likelyParticleID;
+}//Close funtion that return the true trackids that participate the most in showers or tracks  hits
+
+//Function return a vector of true hits of a reco particle
+std::vector<art::Ptr<recob::Hit> > ana::TrackValidation::TrueHitsOfParticle(detinfo::DetectorClocksData const& clockData,
+                                                                            int MotherTrkIds,
+                                                                            std::vector<int>& DaugthersIdsVec,
+                                                                            std::vector<art::Ptr<recob::Hit> > RecoParticleHits){
+  //###### Declaration and Initiate variables #######
+  std::vector<art::Ptr<recob::Hit> >  hitsOfTrkIds;
+  for(auto const& hit : RecoParticleHits){
+    int TrkIdOfHit = FunTrueParticleID(clockData, hit);
+    if(TrkIdOfHit == MotherTrkIds){
+      hitsOfTrkIds.push_back(hit);
+    }//colse condition statement
+    else if(std::find(DaugthersIdsVec.begin(), DaugthersIdsVec.end(), TrkIdOfHit) != DaugthersIdsVec.end()){
+      hitsOfTrkIds.push_back(hit);
+    }//Close else statement
+  }//close looping over all reco hits
+  return hitsOfTrkIds;
+}//Close the function that return a vector of ture hits of a reco particle
+
+//Gets the total energy deposited by a track  by looping through the MC info on each wire.
+std::map<int,float> ana::TrackValidation::TrueEnergyDepositedFromMCTracks(const std::vector<art::Ptr<sim::SimChannel> > &simchannels){
+  //###### Declaration and Initiate variables #######
+  art::ServiceHandle<geo::Geometry> geom;
+  std::map<int,float> total_energies;
+  //
+  // Loop over SimChannel
+  for(auto const& simchannel : simchannels){
+    //
+    //Get the plane.
+    raw::ChannelID_t Channel = simchannel->Channel();
+    std::vector<geo::WireID> Wire = geom->ChannelToWire(Channel);
+    //sbnd is one channel per wire so the vector should be size one.
+    int  PlaneID = (Wire[0]).Plane;
+
+    //Get the TDCIDEMap (Charge vs Time)
+    auto tdc_ide_map = simchannel->TDCIDEMap();
+
+    //Loop through the map
+    for(auto const& tdc_ide_pair : tdc_ide_map){
+      //Get the IDEs associated to the TDC?
+      auto const& ide_v = tdc_ide_pair.second;
+
+      //Loop over the IDEs and add the energy. Only count from the collection plane.
+      for(auto const& ide : ide_v){
+        if(PlaneID == 2){
+          total_energies[std::abs(ide.trackID)] +=  ide.energy;
+        }
+      }
+    }
+  }//Close the loop over the simchannels
+  return total_energies;
+}//Close the function that return the total energy deposited by a track.
+
+//Function remove non contained particles
+void ana::TrackValidation::RemoveNoneContainedParticles(std::map<int,std::vector<int> >&  TracksMothers,
```

Comment:
Surely it is better to query the start and end of the particle rather than rely on the deposited energy

Reply:
resolved